### DISCUSSION TITLE
x-ref undefined on element with x-data

### DISCUSSION THREAD

**Discussion by alterastro** [alpinejs/alpine#3005] [7. Bugs (archived)]
When developing components, I encountered unpredictable x-ref behavior:

```html
<div x-data="Test()">
  <div x-ref="firstRef">
    First
  </div>
  <div x-data="{}" x-ref="secondRef">
    Second
  </div>
</div>
```

When trying to access an element with $refs containing x-ref and x-data on the same element, undefined is returned.
An example can be found at the [link](https://codepen.io/alterastro/pen/BaYXENV).

It seems to me that this is not correct behavior, since we can refer to the ref of a nested Alpine component.
Can anyone say something about this?

**Comment by Martleby** [comment]
I had a look through the source code and it seems deliberate to me that x-refs are scoped to the component they appear in. I guess this make sense: in your example, should secondRef appear on the nested component, the parent component or both? Now imagine a (slightly ludicrous) scenario where you had 30 levels of nested components and the deepest component has a x-ref on it. Should it be available on all of the parent components above it? If it did, I could imagine a complete nightmare of trying to debug errors/bugs/weird unexpected behavior!

I'm guessing you want to do some sort of communication between a parent component and its children? If so I tend to use props and events to do that communication - it's a more flexible and robust way of doing and less prone to errors, I also think it expresses intent better and reduces tight coupling between components.

**Reply by alterastro** [reply]
> 

I disagree about nesting levels and the potential chaos this can lead to. We already have a similar behavior for component properties: if a property is not overridden, then it is available in a nested component.

I agree that events and properties are probably the best way out, but not for all situations. For example, I ran into this problem while wrapping an external library.

**Reply by SimoTod** [reply]
You are trying to do the opposite though, you are trying to access a ref defined on a child component from the parent. This is not allowed even with property (try to define foo in the child component and access foo from the parent, you'll get an error).

**Reply by alterastro** [reply]
Yes you are right. This is my mistake, I often forget about the difference between Vue and Alpine.
The execution context will be different when x-ref is set to an element with x-data.

Thank you clarification!
